network/peer.go

```diff
@@ -836,3 +879,45 @@ func (p *peer) getIP() utils.IPDesc {
 	defer p.ipLock.RUnlock()
 	return p.ip
 }
+
+// addAlias marks that we have found another
+// IP that we can connect to this peer at.
+//
+// assumes stateLock is held
+func (p *peer) addAlias(ip utils.IPDesc) {
+	p.ipLock.Lock()
+	defer p.ipLock.Unlock()
+
+	p.net.peerAliasIPs[ip.String()] = struct{}{}
+	p.aliases = append(p.aliases, alias{
+		ip:    ip,
+		added: p.net.clock.Time(),
+	})
+}
+
+// releaseAliases removes enables network to
+// reconnect on aliases that have timed out. To
+// force all aliases to be removed, the caller
+// should set now to be the zero time.
+func (p *peer) releaseAliases(now time.Time, holdsStateLock bool) {
+	p.ipLock.Lock()
+	defer p.ipLock.Unlock()
+
+	for len(p.aliases) > 0 {
+		next := p.aliases[0]
+		if !now.IsZero() && now.Sub(next.added) < p.net.peerAliasTimeout {
+			return
+		}
+		p.aliases = p.aliases[1:]
+
+		if !holdsStateLock {
```

Comment:
@danlaine I found this was the best way to unify `releaseAliases` without forcing `monitorAliases` to acquire the `stateLock` for the entire duration of the `releaseAliases` call.